src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -121,99 +338,160 @@ private string CreateVaryByCacheKey(CachedVaryBy varyBy)
-            Buffer = new MemoryStream();
-            HttpContext.Response.Body = Buffer;
-        }
+                responseServed = true;
+            }
 
-        internal bool OnResponseStarting()
-        {
-            // Evaluate the response headers, see if we should buffer and cache
-            CacheResponse = true; // TODO:
-            return CacheResponse;
+            return responseServed;
         }
 
-        internal void FinalizeCaching()
+        internal void FinalizeCachingHeaders()
         {
             if (CacheResponse)
             {
+                // Create the cache entry now
                 var response = HttpContext.Response;
-                var varyHeaderValue = response.Headers["Vary"];
+                var varyHeaderValue = response.Headers[HeaderNames.Vary];
+                _cachedResponseValidFor = ResponseCacheControl.SharedMaxAge
+                    ?? ResponseCacheControl.MaxAge
+                    ?? (ResponseHeaders.Expires - _responseTime)
+                    // TODO: Heuristics for expiration?
+                    ?? TimeSpan.FromSeconds(10);
 
                 // Check if any VaryBy rules exist
                 if (!StringValues.IsNullOrEmpty(varyHeaderValue))
                 {
                     var cachedVaryBy = new CachedVaryBy
                     {
                         // Only vary by headers for now
+                        // TODO: VaryBy Encoding
                         Headers = varyHeaderValue
                     };
 
-                    Cache.Set(_cacheKey, cachedVaryBy);
+                    // TODO: Overwrite?
+                    Cache.Set(_cacheKey, cachedVaryBy, _cachedResponseValidFor);
                     _cacheKey = CreateCacheKey(cachedVaryBy);
                 }
 
+                // Ensure date header is set
+                if (ResponseHeaders.Date == null)
+                {
+                    ResponseHeaders.Date = _responseTime;
+                }
+
                 // Store the response to cache
-                var cachedResponse = new CachedResponse
+                _cachedResponse = new CachedResponse
                 {
-                    StatusCode = HttpContext.Response.StatusCode,
-                    Body = Buffer.ToArray()
+                    Created = ResponseHeaders.Date.Value,
+                    StatusCode = HttpContext.Response.StatusCode
                 };
-                foreach (var pair in HttpContext.Response.Headers)
+
+                foreach (var header in ResponseHeaders.Headers)
                 {
-                    cachedResponse.Headers[pair.Key] = pair.Value;
+                    if (header.Key != HeaderNames.Age
+                        && header.Key != HeaderNames.SetCookie)
+                    {
+                        _cachedResponse.Headers.Add(header);
+                    }
                 }
+            }
+            else
+            {
+                UnhookResponseStream();
```

Comment:
OnStarting is only for setting headers, you don't have full control of the response. This could break upstream wrappers.